[wasm] Jiterpreter multithreading support

kg opened the pull request:
This PR enables jiterpreter traces when threading is enabled by making lots of changes:
* We pre-allocate function table entries for each type of jitted code, and fill all those slots with safe placeholder functions that can be called from any thread without crashing. For interp_entry we install the default implementation, for traces we install a function that skips the enter/monitor point without doing anything, and for do_jit_call we install a function that sets a special 'use the slow path' flag. The placeholder for traces is also responsible for ensuring that each thread gets its own compiled copy of traces in order to run them.
* When getting a function pointer for jitted code, we atomically reserve a slot in the pre-allocated tables and install it in that slot on the current thread. Other threads will have the placeholder function in that slot, but when it's called it will automatically JIT the trace again and install it.
* All jiterpreter opcodes were changed to use an 8-byte layout that allows us to patch them using 2-byte atomics:
`JITERPRETER_xxx [2-byte relative function pointer value] [4-byte trace index]`
This means that to run a trace we need to add the value of a global to the function pointer we loaded from memory, but the overhead of that is not too bad. The existing opcode patching model where we reuse the 4-byte storage for different values just was not possible to do in a thread-safe way given WASM's alignment constraints.
* Code that manipulates jiterpreter opcodes now uses a struct that explicitly specifies their layout instead of explicitly doing READ32 and whatnot. This mostly was just to make it easier to update all the code that touched them when I changed the layout.
* The jiterpreter's hit counting logic is updated to also use atomics so that only one thread will ever observe the counter hitting a threshold, and so that other threads shouldn't race to JIT code or patch an opcode. In some cases it seems like the race still happens, but since we use atomics it generates a warning but doesn't crash.
* The jiterpreter's monitoring implementation is updated to use atomics.
* New runtime options are added for configuring the size of the jiterpreter's tables. (For some reason, the interp_entry tables need to be REALLY big). It seems likely that in production we may hit cases where the tables are too small or too big, since there is a limit to how big the function table can get and an AOT'd binary of sufficient size may already have 100k+ table entries.
* All jiterpreter counters are now tracked in C instead of JS and updated using atomics, which allows us to share that data (like number of traces compiled, number of bytes generated) between threads.
* As mentioned above, we JIT a unique copy of the trace for each thread that needs to run it, so we can install it into the function pointer table for that thread. In my testing in Chrome's devtools the unique copies of the traces all have the same hash, which hopefully means that v8 is able to share the compiled code and potentially other data between them. This does mean that we will waste CPU re-jitting (this is fixable, but a lot of work) and waste memory on duplicating traces.
* Traces now take an extra argument (sigh) that contains the address of the trace entry point opcode. This is used so that the placeholder trace function can acquire its own trace index in order to JIT itself on other threads.
* I patched our emscripten libraries to expose a getter for the wasm memory object so that we have access to it in multi-threaded configurations. Existing methods of getting it don't work when threads are active, see https://github.com/emscripten-core/emscripten/issues/19832
* I did some work to try and make the jit_call and interp_entry (AOT) stuff thread safe, but it doesn't work yet, so those are still disabled.
* I changed most of the jiterpreter atomics code to use stdatomic.h because I was having problems with mono_atomic and we are missing some of the atomic ops for 2-byte types. (I later determined that our atomics work fine, it was some other sort of bug on my end). It might be necessary to undo these changes, but we know that the wasm target should always have this header so it's at least safe to use them (and I verified the generated code has the wasm atomics I expect.)
* Trace transfer was removed because making it work with the addition of the new ip argument to traces was nontrivial and I am not sure the optimization is that valuable anymore.

So far I've tested this with the basic browser threading sample (fibonacci) and the threaded version of the raytracer that pavel and aleksey put together.

kg committed:
Checkpoint: Jiterp threading scaffolding

Checkpoint

Checkpoint

Checkpoint thread safety work

Checkpoint

Fix build

Disable traces in threaded builds due to emscripten compat issue

STASH: Debugging atomics issue

Checkpoint threading support

Allocate jiterp tables in each thread's startup

Make jiterpreter counters work cross-thread

Checkpoint

Checkpoint

Checkpoint

Checkpoint

Basic implementation of trace self-healing so all threads will eventually jit all traces

Don't force atomics anymore

Checkpoint

Cleanup

Fix jiterp elapsed time

Disable table allocation message unless stats are active

Fix build

Fix missing linker imports in threaded build
Fix do_jit_call dispatcher attempting to use not-yet-allocated JIT tables
Don't reuse the jit call wrapper table for do_jit_call dispatcher

Raise default table size

Fix typo

Comment out debug print

Fix boost_back_branch_target

Silence unhelpful clang warning

Silence another warning

Address PR feedback: Unify things around Module.getMemory

Revert corebindings changes

kg commented:
There are two problems with enabling traces in threaded builds right now:
* Some emscripten stdlib function(s), presumably atomics, explode for unaligned accesses. I'm not sure what to do about this, since it seems like they probably require 8-byte alignment and opcodes are 2-byte aligned. We might need a global lock for opcode patching, which really sucks.
* Module.asm.memory is not a WebAssembly.Memory in threaded builds for some reason. I opened an emscripten issue to ask how I'm supposed to get at the Memory object.

kg commented:
Almost ready for review. Despite what the comment says, I realized it's not actually fine to use the jitcall table for do_jit_call_indirect, we need special handling for that function

pavelsavara commented on src/mono/wasm/runtime/es6/dotnet.es6.pre.js:
```diff
@@ -1,3 +1,5 @@
 _nativeModuleLoaded = true;
-createDotnetRuntime = Module = createDotnetRuntime(Module);
\ No newline at end of file
+Module["getMemory"] = function () { return wasmMemory; }
```
I would prefer accessing it via `Module.asm.memory`, we already have it here:
https://github.com/dotnet/runtime/blob/4964d542fccea3e80777a62966a7615d98bbb2ea/src/mono/wasm/runtime/startup.ts#L493

Could you please refactor it and move it to memory.ts ?

kg replied:
We can't use Module.asm.memory or Module.wasmMemory. See https://github.com/emscripten-core/emscripten/issues/19832

pavelsavara replied:
Do you mean that the line 493 above doesn't work ?

kg replied:
That's correct, in my testing Module.wasmMemory is undefined and Module.asm.memory is the wrong type to import into another module

pavelsavara replied:
I never tested snapshot in MT. Could you please fix this place to use your new method ? 
Also, could you please add link from `dotnet.es6.pre.js` to https://github.com/emscripten-core/emscripten/pull/19816 so that we could simplify it when we update emsdk.

pavelsavara replied:
thanks a lot!

pavelsavara commented on src/mono/wasm/runtime/jiterpreter-jit-call.ts:
```diff
@@ -281,15 +282,16 @@ export function mono_jiterp_do_jit_call_indirect(
                 },
                 m: {
-                    h: (<any>Module).asm.memory
+                    h: (<any>Module).getMemory()
```
are the cryptic member names really worth it here ?

kg replied:
Yes, they make the module header smaller

kg commented:
I'm not sure what the right solution is for the worker interop exports/imports, I was unable to successfully build locally until I removed all the conditional logic around them - the linker was complaining it couldn't find those two functions. I'm not sure it's worth making them conditional since their implementation is guarded and they're small.

pavelsavara commented on src/mono/wasm/runtime/corebindings.c:
```diff
@@ -67,10 +65,8 @@ void bindings_initialize_internals (void)
 	mono_add_internal_call ("Interop/Runtime::RegisterGCRoot", mono_wasm_register_root);
 	mono_add_internal_call ("Interop/Runtime::DeregisterGCRoot", mono_wasm_deregister_root);
 
-#ifndef DISABLE_THREADS
```
Perhaps these could be reverted now ?

kg replied:
Will test reverting all of them, yeah!

kg commented:
I expect this to cause some small amount of performance regression in single-threaded configurations, so setting NO-MERGE until the perf tooling is all functioning as expected again and we will be able to spot any unexpected changes

pavelsavara commented on src/mono/wasm/runtime/exports-linker.ts:
```diff
@@ -113,6 +109,10 @@ export function export_linker(): any {
         mono_wasm_ends_with,
         mono_wasm_index_of,
 
+        // corebindings.c
```
please move them back to MT

pavelsavara commented on src/mono/wasm/runtime/types/internal.ts:
```diff
@@ -439,8 +439,8 @@ export declare interface EmscriptenModuleInternal {
-    asm: { memory?: WebAssembly.Memory };
-    wasmMemory?: WebAssembly.Memory;
+    asm: any;
+    getMemory(): any;
```
is the return type `WebAssembly.Memory` ?

kg replied:
For asm["memory"] in MT, no. Will find out for getMemory.

kg commented:
Tried reverting the corebindings changes, but when I do I can't run tests locally no matter what I do. Will see if it works on CI.

pavelsavara commented:
I will start reading the code in a minute, but before that, few questions on the description.

>     * The jiterpreter's hit counting logic is updated to also use atomics 
>     * The jiterpreter's monitoring implementation is updated to use atomics.
>     * All jiterpreter counters are now tracked in C instead of JS and updated using atomics

Is all the atomics introduced here only in place for MT build flavor ? Is there perf or size impact on non-MT build ?

>     (For some reason, the interp_entry tables need to be REALLY big). 

Could you please elaborate, I probably don't  have the context to understand what this is about.

>     we will waste CPU re-jitting (this is fixable, but a lot of work) 

We could have "repository" of SharedArrayBuffer blobs by trace index on the main thread and broadcast updates to workers. Probably another PR.

>     * Traces now take an extra argument (sigh) that contains the address of the trace entry point opcode. 

You could install different wrapper to each slot, like `table[currentIndex] = placeholderFn.apply(null, currentIndex)`. That will let you to call it without the extra argument from WASM and the generated trace doesn't need to know about the extra arg.

>     * I patched our emscripten libraries to expose a getter for the wasm memory 
I still want it strongly typed :)

>     * I changed most of the jiterpreter atomics code to use stdatomic.h ... It might be necessary to undo these changes

Was this undone already ?

>     * Trace transfer was removed 

What is "Trace transfer " ?

pavelsavara commented on src/mono/mono/mini/interp/jiterpreter.c:
```diff
@@ -1320,6 +1346,82 @@ mono_jiterp_is_imethod_var_address_taken (InterpMethod *imethod, int offset) {
+	int index = atomic_fetch_add ((atomic_int *)&table->next_index, 1);
+	if (index > table->last_index) {
+		if (index == (table->last_index + 1))
+			g_printf ("MONO_WASM: Jiterpreter table %d is out of space (%d entries allocated)\n", type, index - table->first_index + 1);
```
this should probably disable jiterp going forward, right ?

kg replied:
No, under normal execution the table could fill up. We don't want to disable the entire jiterpreter because one table is full.

pavelsavara commented on src/mono/mono/mini/interp/jiterpreter.c:
```diff
@@ -1187,6 +1170,44 @@ mono_jiterp_get_member_offset (int member) {
+EMSCRIPTEN_KEEPALIVE double
+mono_jiterp_modify_counter (int counter, double delta) {
+	long actual_delta = (long)(delta * JITERP_COUNTER_UNIT);
+	long actual_result = atomic_fetch_add ((atomic_long *)mono_jiterp_get_counter_address (counter), actual_delta);
```
is it worth having `#ifdef DISABLE_THREADS` ?

kg replied:
I thought not, but you're right that it could make the binary larger by pulling in stdatomic implementation in single-thread builds, so I'll ifdef it.

pavelsavara commented on src/mono/mono/mini/interp/jiterpreter.c:
```diff
@@ -21,6 +21,7 @@ void jiterp_preserve_module (void);
 #define jiterp_assert(b)
 #endif
 
+#include <stdatomic.h>
```
revert to mono primitives

pavelsavara commented on src/mono/mono/mini/interp/interp.c:
```diff
@@ -8900,4 +8901,52 @@ mono_jiterp_get_opcode_info (int opcode, int type)
 }
 
+EMSCRIPTEN_KEEPALIVE int
+mono_jiterp_dummy_trace (void *_frame, void *pLocals, JiterpreterCallInfo *cinfo, const guint16 *ip)
```
consider renaming it. Is this the placeholder/stub/trampoline ?

kg replied:
Yeah this is the placeholder. I'll come up with new names if this one isn't obvious.

kg commented:
> I will start reading the code in a minute, but before that, few questions on the description.
> 
> > ```
> > * The jiterpreter's hit counting logic is updated to also use atomics 
> > * The jiterpreter's monitoring implementation is updated to use atomics.
> > * All jiterpreter counters are now tracked in C instead of JS and updated using atomics
> > ```
> 
> Is all the atomics introduced here only in place for MT build flavor ? Is there perf or size impact on non-MT build ?

Right now most of the atomics are not ifdefed out, but size increase is a good reason to do it, so I'll do it.

> 
> > ```
> > (For some reason, the interp_entry tables need to be REALLY big). 
> > ```
> 
> Could you please elaborate, I probably don't have the context to understand what this is about.

We have to pre-reserve tables for each kind of generated code. I haven't been able to figure out why, but the ones for interp_entry - transitions from native code to interpreter - have to be very big. The size I picked actually isn't big enough for System.Text.Json.Tests.

> 
> > ```
> > we will waste CPU re-jitting (this is fixable, but a lot of work) 
> > ```
> 
> We could have "repository" of SharedArrayBuffer blobs by trace index on the main thread and broadcast updates to workers. Probably another PR.

The problem is that we need to reconstruct the correct imports on each thread, so we would need to serialize all of that somewhere while building the trace the first time and stash it. Some constants in the trace could also be thread-local, so we would need to record their locations and patch them.

> 
> > ```
> > * Traces now take an extra argument (sigh) that contains the address of the trace entry point opcode. 
> > ```
> 
> You could install different wrapper to each slot, like `table[currentIndex] = placeholderFn.apply(null, currentIndex)`. That will let you to call it without the extra argument from WASM and the generated trace doesn't need to know about the extra arg.

Yeah, I considered doing this. The problem is that we would now need to either include ~4000 unique dummy functions in the .wasm file (not reasonable) or generate them at startup (also kind of unreasonable)

> 
> > ```
> > * I patched our emscripten libraries to expose a getter for the wasm memory 
> > ```
> 
> I still want it strongly typed :)

Yeah, I need to figure this out. Thanks for the reminder.
> 
> > ```
> > * I changed most of the jiterpreter atomics code to use stdatomic.h ... It might be necessary to undo these changes
> > ```
> 
> Was this undone already ?

No. As I mentioned on discord, I decided to stick with stdatomic because we don't have mono_ versions of all the atomics I need for the types I'm using, and adding new mono_ atomics would require me to implement them on every platform. So I decided jiterpreter.c specifically will only use stdatomic

> 
> > ```
> > * Trace transfer was removed 
> > ```
> 
> What is "Trace transfer " ?

It's an old optimization that predates the jiterp having support for backward branches. It used to be worthwhile but in my testing it doesn't actually do anything anymore, so instead of making it thread-safe I just removed it.

kg commented on src/mono/wasm/runtime/jiterpreter-support.ts:
```diff
@@ -1587,28 +1657,29 @@ export function copyIntoScratchBuffer(src: NativePointer, size: number): NativeP
+        const fnTable = getWasmFunctionTable();
+        fnTable.set(index, f);
+    } else {
+        // mono_log_error(`Table ${table} is full, no space for ${f}`);
```
Unfortunately the table filling up is normal so we have to handle it and continue.

kg commented on src/mono/wasm/runtime/jiterpreter-support.ts:
```diff
@@ -1587,28 +1657,29 @@ export function copyIntoScratchBuffer(src: NativePointer, size: number): NativeP
+    mono_assert (theModule["asm"], "Module['asm'] not available yet");
     if (!wasmTable)
-        wasmTable = (<any>Module)["asm"]["__indirect_function_table"];
+        wasmTable = theModule["asm"]["__indirect_function_table"];
```
Function pointers live in __indirect_function_table, similar to wasmMemory

kg commented on src/mono/wasm/runtime/jiterpreter-support.ts:
```diff
@@ -21,13 +21,94 @@ export declare interface MintOpcodePtr extends NativePointer {
     __brand: "MintOpcodePtr"
 }
 
+export const enum JiterpCounter {
```
Yeah a jiterpreter-enums.ts is a good idea

kg commented on src/mono/mono/utils/options-def.h:
```diff
@@ -74,7 +73,7 @@ DEFINE_BOOL(jiterpreter_interp_entry_enabled, "jiterpreter-interp-entry-enabled"
 #else
 // traces_enabled controls whether the jiterpreter will JIT individual interpreter opcode traces
-DEFINE_BOOL_READONLY(jiterpreter_traces_enabled, "jiterpreter-traces-enabled", FALSE, "JIT interpreter opcode traces into WASM")
+DEFINE_BOOL(jiterpreter_traces_enabled, "jiterpreter-traces-enabled", TRUE, "JIT interpreter opcode traces into WASM")
```
In MT it was previously disabled and couldn't be enabled (read-only), now it's enabled by default, but you can turn it off

kg commented on src/mono/wasm/runtime/jiterpreter-support.ts:
```diff
@@ -2033,3 +2094,43 @@ function updateOptions() {
+    // A partial solution would be to merge the tables based on argument count instead of exact type,
+    //  then create special placeholder functions that examine the rmethod to determine which kind
+    //  of method is being called.
+    const tableSize = options.tableSize, interpEntryTableSize = options.aotTableSize,
```
If you somehow changed these values between startup and creation of the new thread, yes. Realistically speaking it's not possible for that to happen. The jiterpreter will notice that the values don't match and it will assert

kg commented on src/mono/wasm/runtime/jiterpreter.ts:
```diff
@@ -660,6 +661,7 @@ function initialize_builder(builder: WasmBuilder) {
             "frame": WasmValtype.i32,
             "locals": WasmValtype.i32,
             "cinfo": WasmValtype.i32,
+            // FIXME: ip
```
This is dead code that was used by trace transfer, good catch

kg committed:
Don't use stdatomic funcs in single threaded builds since they might make the binary bigger

kg commented on src/mono/wasm/runtime/jiterpreter.ts:
```diff
@@ -1216,8 +1247,8 @@ export function jiterpreter_dump_stats(b?: boolean, concise?: boolean) {
 
-    for (const k in counters.simdFallback)
-        mono_log_info(`// simd ${k}: ${counters.simdFallback[k]} fallback insn(s)`);
+    for (const k in simdFallbackCounters)
```
This is for interpreter SIMD, which is enabled even if wasm SIMD is disabled

pavelsavara replied:
could it be made conditional in interp ?

kg committed:
Remove dead code

kg committed:
Address PR feedback

kg committed:
Address PR feedback

kg committed:
Fix CI build

kg commented:
LibraryTests failure is in hybrid ICU, looks like maybe the browser ICU data doesn't match what we expect it to be

pavelsavara commented:
/azp run runtime-wasm

pavelsavara commented on src/mono/mono/mini/interp/jiterpreter.c:
```diff
@@ -1287,27 +1358,26 @@ mono_jiterp_get_rejected_trace_count ()
+
+	// Atomically update the hit count. We may lose a race here, but that's not a big
+	//  problem since losing the race indicates that the trace entry point is probably hot.
+	mono_atomic_cas_i64 (&trace_info->hit_count, new_hit_count, old_hit_count);
```
#ifdef DISABLE_THREADS

kg commented:
> > > > (For some reason, the interp_entry tables need to be REALLY big).
> > 
> > 
> > We have to pre-reserve tables for each kind of generated code. I haven't been able to figure out why, but the ones for interp_entry - transitions from native code to interpreter - have to be very big. The size I picked actually isn't big enough for System.Text.Json.Tests.
> 
> Do you mean that interp needs to to have many re-entry thunks ? Do they have different signatures ? Do they carry some values in a closure ? Does the number of them grow with number of interpreted methods ? Or with number of jiterp traces ?

Yes, each thunk is dedicated to a specific method. They have different signatures (there's one table per signature) and they close over their target method

> 
> > > We could have "repository" of SharedArrayBuffer blobs
> > 
> > 
> > Some constants in the trace could also be thread-local
> 
> Is that pointer to a local stack ? To instance of something that is thread specific ? Could we have TLS getter for jiterp ?

Yeah some sort of TLS getter is one potential solution here, but it would make things much slower. Another option is to use a global import for the constant and then specify a different value each time we instantiate the module - I did tests with imported globals for constants before, and while it worked it made traces slower, so it's not enabled right now.

The thread local values are things like scratch buffers.

> 
> > > > * Traces now take an extra argument
> > > 
> > > 
> > > You could install different wrapper to each slot
> > 
> > 
> > Yeah, I considered doing this. The problem is that we would now need to either include ~4000 unique dummy functions in the .wasm file (not reasonable) or generate them at startup (also kind of unreasonable)
> 
> I was not proposing wasm methods, I was proposing to create JS wrappers during start. Functions 4000 per thread is not too bad, they would be each executed just once, and therefore not JITed by the browser.
> 
> But I don't know, maybe the extra argument doesn't have significant performance hit and it it's not worth it ?

You can't put JS functions directly into the WASM function table right now. You have to first import them into a WASM module and then re-export them, which is what Module.addFunction does. So we would end up needing to instantiate a bunch of modules in order to create these wrappers. The cost of that might not be too bad, but it's complicated to implement and hard to measure the impact.

> 
> > adding new mono_ atomics would require me to implement them on every platform.
> 
> Why you can't just implement wasm and let the other platforms throw? As the methods are new, they would not be used elsewhere until needed.

I guess I could do that, but it makes me uncomfortable :)

kg committed:
Address PR feedback
Adjust default table size a bit

kg committed:
Update type

kg commented:
Re: earlier questions, `Module.getMemory()` returns a `WebAssembly.Memory` in both ST and MT builds as expected (with `Module.getMemory().buffer` being a `SharedArrayBuffer` in MT builds)
